include/envoy/buffer/buffer.h

```diff
@@ -18,6 +18,29 @@ struct RawSlice {
   size_t len_ = 0;
 };
 
+/**
+ * A wrapper class to facilitate passing in externally owned data to a buffer via addBufferFragment.
+ * When the buffer no longer needs the data passed in through a fragment, it calls done() on it.
+ */
+class BufferFragment {
+public:
+  virtual ~BufferFragment(){};
```

Comment:
Should the destructor be protected?  Calling done() is the way destroy this, so I don't think the destructor should be accessible on the interface.

Reply:
Hmm, calling "done" need not necessarily destroy it.

Reply:
Actually now that I think about it, it wouldn't be possible to bind "this" to the releasor in the constructor initializer would it?

My idea was to have the releasor callback be responsible for deleting the fragment object if necessary. But may not work if the callback has to be passed in to the constructor.

Reply:
I guess my point is that a Buffer should not be permitted to destroy the BufferFragment.  The API contract is that the Buffer will call done() when it's done with it.  Hence a protected destructor.

Reply:
Wouldn't this also disallow the owner of the fragment from destroying it?

Reply:
Actually never mind - this would only apply to the base class's destructor. I didn't quite understand this part.

Reply:
I've gone ahead and done this. Thank you!